File: SpiralMatrix_LC54.java
//Time: O(m*n)

//Space = O(1)


class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> answer = new ArrayList<>();
        
        if(matrix == null || matrix.length == 0 || matrix[0].length == 0)
            return answer;
        
        //initialise pointers
        int R1 = 0; //top row pointer
        int R2 = matrix.length - 1; //bottom row pointer
        int C1 = 0; //left col pointer
        int C2 = matrix[0].length - 1; //right row pointer
        
        //traverse until pointers cross
        while(R1 <= R2 && C1 <= C2){
            //traverse top row, left to right
            for(int col = C1; col <= C2;col++)
                answer.add(matrix[R1][col]);
            R1++;
            
            //traverse last column for row range
            for(int row = R1; row <= R2; row++)
                answer.add(matrix[row][C2]);
            C2--;
            
            //now, may be R1 > R2 and C1 > C2
            if(R1 <= R2 && C1 <= C2){
                //traverse bottom row for all column 
                for(int col = C2; col >= C1; col--)
                    answer.add(matrix[R2][col]);
                R2--;
                
                //traverse 1st column for row range
                for(int row = R2; row >= R1; row--)
                    answer.add(matrix[row][C1]);
                C1++;                
            }
        }
        return answer;        
    }
}

File: spiralMatrix-LC 54.py
# Time Complexity = O(m * n), since the matrix size is m * n
# Space Complexity = O(1)

class Solution:
    def spiralOrder(self, matrix: list[list[int]]) -> list[int]:
        if matrix == None or len(matrix) == 0:
            return []
        
        m = len(matrix) 
        n = len(matrix[0])
        result = [] * (m * n)
        # Initialising the 4 directions (top, bottom, left, right)
        top = 0
        bottom = m - 1
        left = 0
        right = n - 1
        
        # The condition deciding when to stop the spiral/iterations
        while(top <= bottom and left <= right):
            # Top row
            for i in range(left, right+1):
                result.append(matrix[top][i])
                print(result)
            top += 1
            
            # Right column
            # Did not check conditions here coz only the top value was changed above and in the below loop we are iterating from top to bottom, so that condition is getting checked by the for loop itself
            # if (top <= bottom and left <= right):
            for i in range(top, bottom+1):
                result.append(matrix[i][right])    
            right -= 1
        
            # Bottom row
            # Checking the conditions again since while loop checks the conditions after 4 iterations. So if the conditions are not valid in the mid, we will get a wrong answer if we keep iterating keeping the original while loop in mind
            if (top <= bottom and left <= right):
                for i in range(right, left-1, -1):
                    result.append(matrix[bottom][i])
                bottom -= 1
            
            # Left column
            if (top <= bottom and left <= right):
                for i in range(bottom, top-1, -1):
                    result.append(matrix[i][left])
                left += 1
            
        return result